reward/reward_distributor.go

```diff
@@ -51,28 +53,41 @@ func NewRewardDistributor(gh governanceHelper) *RewardDistributor {
 }
 
 // getTotalTxFee returns the total transaction gas fee of the block.
-func (rd *RewardDistributor) getTotalTxFee(header *types.Header, rewardConfig *rewardConfig) *big.Int {
+func (rd *RewardDistributor) getTotalTxFee(header *types.Header, rewardConfig *rewardConfig, isKIP71Forked bool) (*big.Int, error) {
 	totalGasUsed := big.NewInt(0).SetUint64(header.GasUsed)
-	totalTxFee := totalGasUsed.Mul(totalGasUsed, rewardConfig.unitPrice)
-	return totalTxFee
+	var totalTxFee *big.Int
+	if isKIP71Forked {
+		if header.BaseFee == nil {
+			logger.Error("KIP-71 hard forked block should have baseFee", "blockNum", header.Number.Uint64())
+			return nil, errInvalidKIP71Block
+		}
+		totalTxFee = totalGasUsed.Mul(totalGasUsed, header.BaseFee)
+		// burned 50% TxFee
+		return totalTxFee.Div(totalTxFee, big.NewInt(2)), nil
+	} else {
+		return totalGasUsed.Mul(totalGasUsed, rewardConfig.unitPrice), nil
+	}
 }
 
 // MintKLAY mints KLAY and gives the KLAY and the total transaction gas fee to the block proposer.
-func (rd *RewardDistributor) MintKLAY(b BalanceAdder, header *types.Header) error {
+func (rd *RewardDistributor) MintKLAY(b BalanceAdder, header *types.Header, isKIP71Forked bool) error {
```

Comment:
How about instead of adding an additional `isKIP71Forked` parameter to every function (MintKLAY, DistributeBlockReward, getTotalTxFee), we check in the caller like below?
Then we can be sure inside those functions that if header.BaseFee is nil, it is before the fork, and otherwise, it is after the fork.
In this case, old unit tests can be left as-is, and new unit tests will contain header.BaseFee. Please tell me what you think.

```diff
diff --git a/consensus/istanbul/backend/engine.go b/consensus/istanbul/backend/engine.go
index 9ce4f16d..15eb59f3 100644
--- a/consensus/istanbul/backend/engine.go
+++ b/consensus/istanbul/backend/engine.go
@@ -442,6 +442,9 @@ func (sb *backend) Finalize(chain consensus.ChainReader, header *types.Header, s
                        pocAddr = stakingInfo.PoCAddr
                }

+               if isKIP71Forked && header.BaseFee == nil {
+                       return nil, errors.New("Invalide KIP-71 block with no baseFee")
+               }
                if err := sb.rewardDistributor.DistributeBlockReward(state, header, pocAddr, kirAddr); err != nil {
                        return nil, err
                }
```

Reply:
I think the idea is a good way to minimize modifying the existing code. Thanks

Reply:
Refactored. Please check again, Thanks.

Reply:
Thanks! Looks much cleaner now :)